README.md
<h1 align="center">
  <img alt="IDENA ARMer Bash Script - fast idena-go network node deployment for Android" src="https://raw.githubusercontent.com/ltraveler/ltraveler/main/images/IDENA_ARMer_400.png" width="400px"/><br/>
  Idena ARMer Script
</h1>
<p align="center"><b>Bash Script implementation</b> of the <b>Idena network node</b> installation wizard.<br>Install <b>Idena-Go</b> node instance on your mobile device with ARM64 architecture.<br>Simple and user-friendly way to setup your Android device as Idena node server.</p>

<p align="center"><a href="https://github.com/ltraveler/idena-runner/releases/latest" target="_blank"><img src="https://img.shields.io/github/v/release/ltraveler/idena-armer?style=for-the-badge&logo=none" alt="idena runner latest version" /></a>&nbsp;<a href="https://wiki.ubuntu.com/FocalFossa/ReleaseNotes" target="_blank"><img src="https://img.shields.io/badge/Kali-20.04(LTS)+-00ADD8?style=for-the-badge&logo=none" alt="Kali minimum version" /></a>&nbsp;<a href="https://github.com/ltraveler/idena-runner/blob/main/CHANGELOG.md" target="_blank"><img src="https://img.shields.io/badge/Build-Stable-success?style=for-the-badge&logo=none" alt="idena-go latest release" /></a>&nbsp;<a href="https://www.gnu.org/licenses/quick-guide-gplv3.html" target="_blank"><img src="https://img.shields.io/badge/license-GPL3.0-red?style=for-the-badge&logo=none" alt="license" /></a>&nbsp;<a href="https://github.com/ltraveler/idena-armer/blob/master/README.ru-RU.md" target="_blank"><img src="https://img.shields.io/badge/readme-РУССКИЙ-orange?style=for-the-badge&logo=none" alt="Скрипт Idena ARMer" /></a></p>

## 🚀&nbsp; Summary
1. Installing **F-Droid** - an alternative app repository.
2. From **F-Droid** installing **Termux** and **Termux:boot**.
3. Running **Termux** and **Termux:boot** for the 1st time to make basic initialization process of those apps.
4. `termux-setup-storage` - giving to Termux required storage permission.
5. Adding **Termux** and **Termux:boot** to the Power Monitor exclusions.
6. `pkg upgrade -y` - updating termux packages to the latest version.
7. Downloading and running `termux_init.sh` script to prepare Termux terminal to host Kali Linux:
```
curl -O https://raw.githubusercontent.com/ltraveler/idena-armer/master/termux_init.sh
chmod +x ./termux_init.sh && ./termux_init.sh
```
8. `./start-kali.sh` - running Kali Linux.
9. `apt update && apt upgrade -y` - updating Kali Linux packages to the latest version.
10. `apt install git -y` - installing git package.
11. `git clone https://github.com/ltraveler/idena-armer.git` - cloning IDENA ARMer repository.
12. `cd idena-armer && chmod +x ARMer_init.sh &&./ARMer_init.sh` - running IDENA ARMer and following the script's installation instructions.
13. Reboot the phone.

## ⏳&nbsp; Update IDENA ARMer to the latest version:
**_All commands must be run inside Kali Linux!_**
1. `cd ~` - returning to the home directory.
2. `rm -rf idena-armer` - removing previously cloned repository.
3. `git clone https://github.com/ltraveler/idena-armer.git` - cloning latest IDENA ARMer repository.
4. `cd idena-armer && chmod +x ARMer_init.sh &&./ARMer_init.sh` - running IDENA ARMer and installing idena-go node client with <i><b>the same username</b></i> that was used for the first-time installation.

## 🐼&nbsp; Xiaomi devices based on MIUI
<b>MIUI battery usage optimization is extremely high</b>. As a result, termux application cannot being run continiously as background process. There is a strong probability that you would need to install a custom ROM. LineageOS is the best one so far for Xiaomi devices based on MIUI OS.

## 📗&nbsp; The detailed installation guide has been published here:
https://medium.com/@idna.project/b9229c010440

## 👀&nbsp; [IDENA Coacher](https://github.com/ltraveler/idena-coacher) — Node Management Tool
1. **IDENA ARMer** will set **IDENA Coacher** to manage your node.
2. It is placed in the home folder of the user that has been used to install idena.
3. Run the script.
```
idena_coacher.sh
```
4. You should see the main window of the node management tool.

<p align="center"><br>
  <img alt="Idena Coacher Node Monitor Tool" Title="IDENA Coacher - User Interface" src="https://raw.githubusercontent.com/ltraveler/ltraveler/main/images/IDENA_Coacher_Monitor_Tool_UI.jpg">
</p>


CHANGELOG.md
## 0.1.6 (Nov 03, 2022)

### Changes

- IDENA Coacher home directory has been added to '$PATH'

## 0.1.5 (Oct 29, 2022)

### Changes

- Additional protection layer has been added to prevent to start idena-go node client during the idena blockchain database rebuilding process.

## 0.1.4 (Oct 29, 2022)

### Bugfix

- Self-checking script have had the wrong call to rebuild `idenachain.db`.

## 0.1.3 (Oct 28, 2022)

### Changes

- First released stable version. If you have an older version, it is highly recommended backup your private key, remove Kali Linux and start over as it is described [here](https://medium.com/@idna.project/b9229c010440#cde3).


termux_init.sh
#!/data/data/com.termux/files/usr/bin/sh

#ASCII color assign
RED="\033[0;31m"
LRED="\033[1;31m"
YELLOW="\033[0;33m"
LYELLOW="\033[1;33m"
GREEN="\033[0;32m"
LGREEN="\033[1;32m"
BLUE="\033[0;34m"
LBLUE="\033[1;34m"
PURPLE="\033[0;35m"
LPURPLE="\033[1;35m"
CYAN="\033[0;36m"
LCYAN="\033[1;36m"
NC="\033[0m" # No Color

pkg upgrade -y
pkg install termux-services -y
pkg install openssh -y
pkg install wget -y
pkg install nano -y
pkg install cronie -y
crond
sv-enable crond
passwd
sshd
mkdir -p /data/data/com.termux/files/home/.termux/boot

cd /data/data/com.termux/files/home/.termux/boot
wget -O start-sshd https://raw.githubusercontent.com/ltraveler/idena-armer/master/.termux/boot/start-sshd
wget -O start-trmx https://raw.githubusercontent.com/ltraveler/idena-armer/master/.termux/boot/start-trmx
wget -O start-tserv https://raw.githubusercontent.com/ltraveler/idena-armer/master/.termux/boot/start-tserv
chmod +x *

cd /data/data/com.termux/files/home/

ifconfig | grep "inet " | awk -F'[: ]+' '{ print $3 }'
echo -e "\n\n${LRED}Please save your IP address from the above list.\n${LYELLOW}Your port number for ${LGREEN}SSH${LYELLOW} has been set to ${LGREEN}8022${LYELLOW}.\nAfter establishing ${LGREEN}SSH connection${LYELLOW}, please keep ${LGREEN}username${LYELLOW} field empty and use the previously set ${LGREEN}password${LYELLOW} to login."
printf "${LRED}Press any key to continue...${NC}"
read ans

pkg install wget openssl-tool proot -y && hash -r && wget https://raw.githubusercontent.com/EXALAB/AnLinux-Resources/master/Scripts/Installer/Kali/kali.sh && bash kali.sh

touch /data/data/com.termux/files/home/.bashrc
grep -q 'bash /data/data/com.termux/files/home/start-kali.sh' '/data/data/com.termux/files/home/.bashrc' || echo "bash /data/data/com.termux/files/home/start-kali.sh" >> /data/data/com.termux/files/home/.bashrc
grep -q '# allow-external-apps = true' '/data/data/com.termux/files/home/.termux/termux.properties' | sed -i '/allow-external-apps/s/^#//g' /data/data/com.termux/files/home/.termux/termux.properties

exit


armer_update.sh
#!/usr/bin/bash
ARMER_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
username=$(echo "$ARMER_DIR" | awk -F\/ '{print $3}')
if [ -d "/home/$username/idena-go" ] 
then
	service idena stop
else
	echo "Error: IDENA ARMer is not installed."
	exit
fi

cd /home/$username/idena-go/
read -p "Would you like me to compile idena-go bin file from the source code?" -n 1 -r
echo "Let's do some updates..." 
if [[ $REPLY =~ ^[Yy]$ ]]
then
        if ! command -v go &> /dev/null
        then
                echo "Compiling idena-go..."
                wget -O go1.17.13.linux-arm64.tar.gz https://go.dev/dl/go1.17.13.linux-arm64.tar.gz
                tar -xvf go1.17.13.linux-arm64.tar.gz
                rm go1.17.13.linux-arm64.tar.gz
                fi

        GOROOT=/home/$username/idena-go/go
        GOPATH=/home/$username/idena-go/go
        PATH=$GOPATH/bin:$PATH

        gover=$(curl -s https://api.github.com/repos/idena-network/idena-go/releases/latest | grep -Po '"tag_name":.*?[^\\]",' | sed 's/"tag_name": "v//g' |  sed 's/",//g')
        wget -O v$gover.tar.gz https://github.com/idena-network/idena-go/archive/refs/tags/v$gover.tar.gz
        tar -xzf v$gover.tar.gz && rm v$gover.tar.gz
        cd idena-go-$gover

        go build -ldflags "-X main.version=$gover" && chmod +x idena-go && cp -r idena-go /home/$username/idena-go/ && cd /home/$username/idena-go && rm -rf ./idena-go-$gover ./go
	service idena start
else
        wget -O idena-go https://github.com/ltraveler/idena-go-arm64/raw/main/idena-go && chmod +x idena-go
	service idena start
fi



ARMer_init.sh
#!/bin/bash 
#
#ASCII color assign
RED="\033[0;31m"
LRED="\033[1;31m"
YELLOW="\033[0;33m"
LYELLOW="\033[1;33m"
GREEN="\033[0;32m"
LGREEN="\033[1;32m"
BLUE="\033[0;34m"
LBLUE="\033[1;34m"
PURPLE="\033[0;35m"
LPURPLE="\033[1;35m"
CYAN="\033[0;36m"
LCYAN="\033[1;36m"
NC="\033[0m" # No Color

set -o pipefail -o nounset -o noclobber -o allexport
ARMER_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )

#Installing all required packages
apt update
apt upgrade -y

reqpkgs=('whiptail' 'moreutils' 'original-awk' 'systemd-cron' 'cron' 'procps' 'gettext' 'jq' 'gcc' 'sudo' 'git' 'curl' 'wget' 'nano' 'screen' 'psmisc' 'unzip' 'htop')
for x in "${reqpkgs[@]}"; do
dpkg -s "$x" &> /dev/null
if [ $? != 0 ]; then
    echo -e "${LRED}Package $x is not instlled. Installing...${NC}"
    apt-get install -y $x; 
#
fi
done

#Adding new user for idena-go client
echo -e "${LYELLOW}Please enter a ${LRED}username${NC} and ${LRED}password${LYELLOW} that you would like to use for this ${LGREEN}Idena Node Daemon${LYELLOW} instance.${NC}"
while
echo -e "${NC}Please do not use ${LGREEN}root${NC} as a username"
read -p "Enter username : " username
[[ $username = 'root' ]] || [[ -z $username ]]
do true; done

#Some preliminary check
if screen -ls | grep -q "$username"; then
    echo "The given username $username has a running idena-go client.\nTerminating..."
    service idena stop
fi


#If user exists we have to save our beans
if id "$username" >/dev/null 2>&1; then
        echo "User $username exists. All keys are saved in ./armer_backup directory"
	mkdir -p /home/$username/ARMer_backup/
	cp -r --backup=t /home/$username/idena-go/datadir/api.key /home/$username/idena-go/datadir/keystore/nodekey /home/$username/ARMer_backup/ 
	rm -rf /home/$username/idena-go
	rm -rf /home/$username/idena-coacher
	rm -rf /home/$username/scripts
	rm -rf /home/$username/flags
	rm -f /etc/init.d/idena
	crontab -l | grep -v 'idena_scrchk.sh' | crontab -
	crontab -l | grep -v 'idena_selfcheck.sh' | crontab -
else
        echo "Creating a user $username..."
	password="$username"
	salt=$(cat /dev/urandom | tr -dc A-Za-z0-9/_- | head -c16)
	pass=$(openssl passwd -5 -salt "$salt" "$password")
	useradd -s /bin/bash -m -p "$pass" "$username"
fi

mkdir /home/$username/idena-go
cd /home/$username/idena-go

read -p "Would you like me to compile idena-go bin file from the source code? " -n 1 -r
echo  "Let's do some updates" 
if [[ $REPLY =~ ^[Yy]$ ]]
then
        if ! command -v go &> /dev/null
        then
                echo "Compiling idena-go..."
                wget -O go1.17.13.linux-arm64.tar.gz https://go.dev/dl/go1.17.13.linux-arm64.tar.gz
                tar -xvf go1.17.13.linux-arm64.tar.gz
                rm go1.17.13.linux-arm64.tar.gz
                fi

        GOROOT=/home/$username/idena-go/go
        GOPATH=/home/$username/idena-go/go
        PATH=$GOPATH/bin:$PATH

        gover=$(curl -s https://api.github.com/repos/idena-network/idena-go/releases/latest | grep -Po '"tag_name":.*?[^\\]",' | sed 's/"tag_name": "v//g' |  sed 's/",//g')
        wget -O v$gover.tar.gz https://github.com/idena-network/idena-go/archive/refs/tags/v$gover.tar.gz
        tar -xzf v$gover.tar.gz && rm v$gover.tar.gz
        cd idena-go-$gover

        go build -ldflags "-X main.version=$gover" && chmod +x idena-go && cp -r idena-go /home/$username/idena-go/ && cd /home/$username/idena-go && rm -rf ./idena-go-$gover ./go
else
        wget -O idena-go https://github.com/ltraveler/idena-go-arm64/raw/main/idena-go
fi

chmod +x idena-go
mkdir -p /home/$username/idena-go/datadir/idenachain.db
cd /home/$username/idena-go/datadir/idenachain.db
wget -O main.zip https://github.com/ltraveler/idenachain.db/archive/refs/heads/main.zip
unzip -jo main.zip
rm main.zip
cd /home/$username/idena-go
wget -O config.json https://raw.githubusercontent.com/ltraveler/idena-runner/main/config.json
#chown -R $username:$username /home/$username/idena-go

screen -S idena -dm /home/$username/idena-go/idena-go --config=/home/$username/idena-go/config.json --datadir=/home/$username/idena-go/datadir

until [ -s /home/$username/idena-go/datadir/api.key ]
do
     sleep 5
done
until [ -s /home/$username/idena-go/datadir/keystore/nodekey ]
do
     sleep 5
done
screen -S idena -X quit

#Private rollback segment
apikey=$( cat /home/$username/idena-go/datadir/api.key )
echo -e ${LBLUE}Your IDENA-node API key is: ${YELLOW}$apikey
prvkey=$( cat /home/$username/idena-go/datadir/keystore/nodekey )
echo -e ${LBLUE}Your IDENA-node PRIVATE key is: ${YELLOW}$prvkey${NC}

while true; do
    read -p "$(echo -e ${LYELLOW}Would you like to add your own ${LGREEN}IDENA NODE PRIVATE KEY${LYELLOW} \(aka ${LRED}nodekey${LYELLOW}\)?$'\n'${LGREEN}Default path: ${LRED}../idena-go/datadir/keystore/nodekey ${LGREEN}[y/N]${NC} )" yn
    case $yn in
        [Yy]* ) read -p "$(echo -e ${LYELLOW}Please enter your IDENA private key:${NC} )"; echo "$REPLY" >| /home/$username/idena-go/datadir/keystore/nodekey; break;;
        [Nn]* ) break;;
        * ) echo "Please answer yes or no.";;
    esac
done

while true; do
    read -p "$(echo -e ${LYELLOW}Would you like to add your own ${LGREEN}IDENA NODE API KEY${LYELLOW} \(aka ${LRED}api.key${LYELLOW}\)?$'\n'${LGREEN}Default path: ${LRED}../idena-go/datadir/api.key  ${LGREEN}[y/N]${NC} )" yn
    case $yn in
        [Yy]* ) read -p "$(echo -e ${LYELLOW}Please enter your IDENA node key: ${NC})"; echo "$REPLY" >| /home/$username/idena-go/datadir/api.key; break;;
        [Nn]* ) break;;
        * ) echo "Please answer yes or no.";;
    esac
done

#System curtain-raiser

cd $ARMER_DIR/armer/ 
envsubst < idena >| /etc/init.d/idena
chmod +x /etc/init.d/idena
update-rc.d idena defaults

#Measure thrice and cut once
tail --lines=+2 idena_bash.sh | envsubst >| idenabash.tmp 
awk 'FNR == NR { lines[$0] = 1; next } ! ($0 in lines) {print}' /root/.bashrc idenabash.tmp | sponge -a /root/.bashrc
rm idenabash.tmp 
#
#idenachain.db health check
mkdir /home/$username/scripts
mkdir -p /home/$username/flags
envsubst < idena_selfcheck.sh >| /home/$username/scripts/idena_selfcheck.sh
# idena-go manual update
cp ../armer_update.sh /home/$username/scripts/
cp ./idena_scrchk.sh /home/$username/scripts/
chmod +x /home/$username/scripts/*.sh
cd /home/$username/scripts
crontab -u root -l >| idenacron
#Particular error cases pre-check
#grep "\*\/30 \* \* \* \* \/home\/$username\/scripts\/idena_selfcheck.sh" idenacron || echo "*/30 * * * * /home/$username/scripts/idena_selfcheck.sh" >> idenacron
echo "*/30 * * * * /home/$username/scripts/idena_selfcheck.sh" >> idenacron
#idena-go daemon run check
#grep "\*\/13 \* \* \* \* \/home\/$username\/scripts\/idena_scrchk.sh" idenacron || echo "*/13 * * * * /home/$username/scripts/idena_scrchk.sh" >> idenacron
echo "*/13 * * * * /home/$username/scripts/idena_scrchk.sh" >> idenacron
crontab -u root idenacron
rm idenacron
/etc/init.d/cron restart
cd $ARMER_DIR
service idena start
#IDENA Coacher installation
cd /home/$username
git clone https://github.com/ltraveler/idena-coacher.git
cd idena-coacher
chmod +x idena_coacher.sh
# Installation has been successfully completed
echo -e "${LRED}IDENA NODE HAS BEEN SUCCESSFULLY INSTALLED" 
echo -e "${LGREEN}FOR IDENA DONATIONS:${NC} 0xf041640788910fc89a211cd5bcbf518f4f14d831"
echo -e "${YELLOW}CONTACT AUTHOR:${NC} ltraveler@protonmail.com"
echo -e "${LBLUE}IDENA PERSONALIZED SHARED NODE SERVICE:${NC} https://t.me/ltrvlr"
exit

